shakedown/dcos/__init__.py

```diff
@@ -1,9 +1,107 @@
 import os
-import dcos
-import dcos.cluster
 import sys
 
+import select
+import threading
+
+import dcos
+import dcos.cluster
 import shakedown
+from .helpers import validate_key, try_close, get_transport, start_transport
+
+
+class TransportManager(object):
+    
+    def __init__(self):
+        self.connections = {}
+        self.mutex = threading.RLock()
+    
+    @staticmethod
+    def key_name(host, username):
+        return "{h}-{u}".format(h=host, u=username)
+    
+    @staticmethod
+    def _check_username(username):
+        return shakedown.cli.ssh_user if not username else username
+
+    def _open_transport(self, host, username, key_path):
+        """Open a new SSH transport/connection to host. This operation
+        is heavy, as it includes authenticating.
+
+        :param host: host or IP of the machine
+        :type host: str
+        :param username: SSH username
+        :type username: str
+        :param key_path: path to the SSH private key for SSH auth
+        :type key_path: str
+        :param noisy: verbose output
+        :type noisy: bool
+        :return: new, connected transport or None
+        """
+        username = self._check_username(username)
+        if not key_path:
+            key_path = shakedown.cli.ssh_key_file
+        
+        key = validate_key(key_path)
+        transport = get_transport(host, username, key)
+    
+        if transport:
+            transport = start_transport(transport, username, key)
+        else:
+            print("error: unable to connect to {}".format(host))
+            return None
+        
+        if transport.is_authenticated():
+            needle = self.key_name(host, username)
+            with self.mutex:
+                # make extra sure nothing else added this needle.
+                # if it did, use the existing one and close this one.
+                close_this_one = needle in self.connections
+                if not close_this_one:
+                    self.connections[needle] = transport
+                returnable = self.connections[needle]
+            
+            if close_this_one:
+                try_close(transport)
+            
+            # scope is weird
```

Comment:
this comment doesn't make sense... `scope is weird`?

Reply:
I will remove this comment.